src/options.rs

```diff
@@ -0,0 +1,180 @@
+        self.rows = rows as c_int;
+    }
+
+    pub fn set_cols(&mut self, cols: u32) {
+        self.cols = cols as c_int;
+    }
+
+    pub fn set_chain_length(&mut self, chain_length: u32) {
+        self.chain_length = chain_length as c_int;
+    }
+
+    pub fn set_parallel(&mut self, parallel: u32) {
+        self.parallel = parallel as c_int;
+    }
+
+    pub fn set_pwm_bits(&mut self, pwm_bits: u8) -> LedMatrixOptionsResult {
+        if pwm_bits > 11 {
+            Err("Pwm bits can only have value between 0 and 11 inclusive")
+        } else {
+            self.pwm_bits = pwm_bits as c_int;
+            Ok(())
+        }
+    }
+
+    pub fn set_pwm_lsb_nanoseconds(&mut self, pwm_lsb_nanoseconds: u32) {
+        self.pwm_lsb_nanoseconds = pwm_lsb_nanoseconds as c_int;
+    }
+
+    pub fn set_brightness(&mut self, brightness: u8) -> LedMatrixOptionsResult {
+        if brightness > 100 || brightness < 1 {
+            Err("Brigthness can only have value between 1 and 100 inclusive")
+        } else {
+            self.brightness = brightness as c_int;
+            Ok(())
+        }
+    }
+
+    pub fn set_scan_mode(&mut self, scan_mode: u32) {
+        self.scan_mode = scan_mode as c_int;
+    }
+
+    pub fn set_led_rgb_sequence(&mut self, sequence: &str) {
+        unsafe {
+            let _ = CString::from_raw(self.led_rgb_sequence);
+            self.led_rgb_sequence = CString::new(sequence).unwrap().into_raw();
+        }
+    }
+
+    pub fn set_pixel_mapper_config(&mut self, mapper: &str) {
+        unsafe {
+            let _ = CString::from_raw(self.pixel_mapper_config);
+            self.pixel_mapper_config = CString::new(mapper).unwrap().into_raw();
+        }
+    }
+
+    pub fn set_hardware_pulsing(&mut self, enable: bool) {
+        if enable {
+            self.disable_hardware_pulsing = 0;
+        } else {
+            self.disable_hardware_pulsing = 1;
+        }
+    }
+
+    pub fn set_refresh_rate(&mut self, enable: bool) {
+        if enable {
+            self.show_refresh_rate = 1;
+        } else {
+            self.show_refresh_rate = 0;
+        }
+    }
+
+    pub fn set_inverse_colors(&mut self, enable: bool) {
+        if enable {
+            self.inverse_colors = 1;
+        } else {
+            self.inverse_colors = 0;
+        }
+    }
+
+    pub fn set_multiplexing(&mut self, multiplexing: u32) {
+        self.multiplexing = multiplexing as c_int;
+    }
+
+    pub fn set_row_addr_type(&mut self, row_addr_type: u32) {
+        self.row_address_type = row_addr_type as c_int;
+    }
+
+    pub fn set_limit_refresh(&mut self, limit_refresh: u32) {
+        self.limit_refresh_rate_hz = limit_refresh as c_int;
+    }
+
+    pub fn set_pwm_dither_bits(&mut self, pwm_dither_bits: u32) {
+        self.pwm_dither_bits = pwm_dither_bits as c_int;
+    }
+
+    pub fn set_panel_type(&mut self, panel_type: &str) {
+        unsafe {
+            let _ = CString::from_raw(self.panel_type);
```

Comment:
whats going on here?

Reply:
We need to use `CString::into_raw("string literal")` when passing the string into the C API, which makes it so rust won't deallocate the string rust made. For it to be cleaned up, we need to use `from_raw` to consume and clean it up. [from_raw docs](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_raw)